### ISSUE TITLE
Azure devops hosted image 20210411.1 builds SSIS packages that cannot be deployed anymore [regression]

### ISSUE THREAD

**Issue by ibruynin** [actions/virtual-environments#3186]
**Description**  
_Introduction_
We have an azure devops build pipeline that builds SSIS ispac artifact on a VS 2017 hosted agent.
We have an azure devops release pipeline that deploys those ispac artifacts to our on-prem SQL database(s) with an on prem agent.

_Issue_

- ispac artifacts built with '20210329.1' can be deployed without any issues.
- ispac artifacts (even rebuilding the same codebase) with '20210411.1' fail with error "Cannot find type Microsoft.SqlServer.Management.IntegrationServices: verify that the assembly...

re-releasing older artifacts (built on '20210329.1') works fine.
So it's clearly a regression issue on the newer version of the image of the hosted agent.

See also https://twitter.com/IBruyninckx/status/1382294948389789696?s=20

I cannot find obvious differences in https://github.com/actions/virtual-environments/blob/win16/20210329.1/images/win/Windows2016-Readme.md vs https://github.com/actions/virtual-environments/blob/win16/20210411.1/images/win/Windows2016-Readme.md

**Area for Triage**:  

**Question, Bug, or Feature?**:  

**Virtual environments affected**  
- [ ] Ubuntu 16.04
- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11.0
- [x] Windows Server 2016 R2
- [ ] Windows Server 2019

**Image version**

This is part of the detailed output in the devops build pipeline:

_Agent job 1_
Pool: Azure Pipelines
Image: VS2017
Agent: Hosted Agent

_Initialize job_
Starting: Initialize job
Agent name: 'Hosted Agent'
Agent machine name: 'fv-az177-900'
Current agent version: '2.184.2'
Operating System
Virtual Environment
Current image version: '20210411.1'
Agent running as: 'VssAdministrator'

**Expected behavior**  
SSIS ispac packages can be deployed regardless of the image being used.

**Actual behavior**  
SSIS ispac packages cannot be deployed when built on the latest version of the image.

**Repro steps**  
1. build an SSIS package (a) on '20210329.1' and deploy on SQL 12.0.5590.1
2. build an SSIS package (a) on '20210411.1' and deploy on SQL 12.0.5590.1

(a) building SSIS packages with the following command line:
```
devenv.com $(build.sourcesdirectory)\<ssdtsolution>.sln /build $(BuildConfiguration)
```

(b) deploy ispac with the following powershell:
```
# Load the IntegrationServices assembly
$loadStatus = [System.Reflection.Assembly]::Load("Microsoft.SQLServer.Management.IntegrationServices, "+
    "Version=$SQLVersion.0.0.0, Culture=neutral, PublicKeyToken=89845dcd8080cc91, processorArchitecture=MSIL")

# Create a connection to the server
$sqlConnectionString = `
    "Data Source=" + $TargetServerName + ";Initial Catalog=master;Integrated Security=SSPI;"
$sqlConnection = New-Object System.Data.SqlClient.SqlConnection $sqlConnectionString

# Create the Integration Services object
$integrationServices = New-Object "Microsoft.SqlServer.Management.IntegrationServices" $sqlConnection

# Get the Integration Services catalog
$catalog = $integrationServices.Catalogs["SSISDB"]

# Create object to the (new) folder
$Folder = $Catalog.Folders[$TargetFolderName]

# Check if folder already exists
if (-not $Folder)
{
    # Folder doesn't exists, so create the new folder.
    Write-Host "Creating new folder" $TargetFolderName
    New-Object "Microsoft.SqlServer.Management.IntegrationServices.CatalogFolder" ($Catalog, $TargetFolderName, $TargetFolderName) -OutVariable Folder
    $Folder.Create()
}
else
{
    Write-Host "Folder" $EnvironmentFolderName "found"
}

Write-Host "Deploying " $ProjectName " project ..."

# Read the project file and deploy it
[byte[]] $projectFile = [System.IO.File]::ReadAllBytes($ProjectFilePath)
$Folder.DeployProject($ProjectName, $projectFile)

Write-Host "Done."
```
Basically, as documented here: https://docs.microsoft.com/en-us/sql/integration-services/ssis-quickstart-deploy-powershell?view=sql-server-ver15#powershell-script

**Comment by Darleev** [comment]
Hello @ibruynin,
Thank for contacting us.
The list of changes can be found in the following PRs:
[Windows 2016](https://github.com/actions/virtual-environments/pull/3157) 
[Windows 2019](https://github.com/actions/virtual-environments/pull/3156)
We are investigating the issue.

**Comment by ibruynin** [comment]
Hi @Darleev

I compared the ispac artifacts in a binary fashion and noticed that they are identical, so I needed to dig a little deeper...

The issue was related to parallel changes in the deployment powershell script (that was also part of the build artifact)
That's why previous builds worked and newer builds did not.

You can close this one and lots of excuses for the inconvenience!

**Comment by dsame** [comment]
@ibruynin thanks for your note! Feel free to re-open this issue or create another one in case if the problem still exists

**Comment by fmms** [comment]
@ibruynin are you hitby #8174 as well? The SSIS DLL versions in the GAC on the agent are very old (2012 - 2016). The recent ones (2017 - 2022) are missing ....
